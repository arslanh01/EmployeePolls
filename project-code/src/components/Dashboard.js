import { connect } from "react-redux";
import Nav from "./Nav";
import Question from "./Question";

const Dashboard = (props) => {
  const unansweredQuestionIds = props.answeredQuestions
    ? Object.values(props.questions)
        .filter((question) => !props.answeredQuestions.includes(question.id))
        .sort((a, b) => b.timestamp - a.timestamp)
    : null;
  return (
    <div>
      <div>
        <Nav />
        <div className="center">
          <img src={props.user.avatarURL} alt="Avatar" className="avatar" />
          <h3>{props.user.name}</h3>
          <div>
            <h3>Unanswered Questions</h3>
            {unansweredQuestionIds.map((question) => {
              return <Question key={question.id} id={question.id} />;
            })}
          </div>
          <div>
            <h3>Answered Questions</h3>
            {props.answeredQuestions.map((id) => {
              return <Question key={id} id={id} />;
            })}
          </div>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = ({ loginUser, users, questions }) => {
  const user = users[loginUser];
  const answeredQuestions = user
    ? Object.keys(users[loginUser].answers).sort(
        (a, b) => questions[b].timestamp - questions[a].timestamp
      )
    : null;
  return {
    user,
    answeredQuestions,
    questions,
  };
};
export default connect(mapStateToProps)(Dashboard);
