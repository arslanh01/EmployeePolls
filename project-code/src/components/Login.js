import { useState } from "react";
import { connect } from "react-redux";
import { setLoginUser } from "../actions/loginUser";

const Login = (props) => {
  console.log(props);
  const [user, setUser] = useState();
  const handleChange = (e) => {
    setUser(e.target.value);
  };
  const handleLogin = () => {
    props.dispatch(setLoginUser(user));
  };
  return (
    <div>
      <select onChange={handleChange}>
        {props.users
          ? props.users.map((user) => {
              return (
                <option key={user} value={user}>
                  {user}
                </option>
              );
            })
          : ""}
      </select>
      <button onClick={handleLogin}>Login</button>
    </div>
  );
};

const mapStateToProps = ({ users }) => {
  return {
    users: Object.keys(users),
  };
};

export default connect(mapStateToProps)(Login);
