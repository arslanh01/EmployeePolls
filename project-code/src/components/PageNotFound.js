import { Link } from "react-router-dom";

const PageNotFound = () => {
  return (
    <div>
      <Link to="/">Go back to dashboard</Link>
      <div className="center">
        <h3>
          Seems like the page you're looking for does not exist or has been
          moved
        </h3>
      </div>
    </div>
  );
};

export default PageNotFound;
