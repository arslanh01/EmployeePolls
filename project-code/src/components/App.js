import { Routes, Route } from "react-router-dom";
import { handleInitialData } from "../actions/shared";
import LoadingBar from "react-redux-loading-bar";
import { useEffect } from "react";
import { connect } from "react-redux";
import Dashboard from "./Dashboard";
import Login from "./Login";
import Poll from "./Poll";

const App = (props) => {
  useEffect(() => {
    props.dispatch(handleInitialData());
  }, [props]);

  return (
    <div>
      <LoadingBar />
      {props.loginUser ? (
        <Routes>
          <Route exact path="/" element={<Dashboard />} />
          <Route exact path="/questions/:question_id" element={<Poll />} />
        </Routes>
      ) : (
        <Login />
      )}
    </div>
  );
};

const mapStateToProps = ({ loginUser }) => ({
  loading: loginUser === null,
  loginUser,
});

export default connect(mapStateToProps)(App);
