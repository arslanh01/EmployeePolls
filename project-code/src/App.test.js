import { render, fireEvent } from "@testing-library/react";
import { store } from "./index";
import { _saveQuestion, _saveQuestionAnswer } from "../src/utils/_DATA";
import { Provider } from "react-redux";
import { BrowserRouter, MemoryRouter } from "react-router-dom";
import Dashboard from "./components/Dashboard";
import App from "./components/App";
import Login from "./components/Login";
import PageNotFound from "./components/PageNotFound";

describe("_saveQuestion tests", () => {
  it("test 1", async () => {
    const question = {};
    question.optionOneText = "Go to movies";
    question.optionTwoText = "Go to walk";
    question.author = "sarahedo";
    const result = await _saveQuestion(question);
    expect(result.author).toEqual("sarahedo");
  });
  it("test 2", async () => {
    const question = {};
    question.optionOneText = "Go to movies";
    await expect(_saveQuestion(question)).rejects.toEqual(
      "Please provide optionOneText, optionTwoText, and author"
    );
  });
});

describe("_saveQuestionAnswer tests", () => {
  it("test 1", async () => {
    const result = await _saveQuestionAnswer({
      authedUser: "sarahedo",
      qid: "vthrdm985a262al8qx3do",
      answer: "optionOne",
    });
    expect(result).toEqual(true);
  });
  it("test 2", async () => {
    await expect(
      _saveQuestionAnswer({
        authedUser: "sarahedo",
        answer: "optionOne",
      })
    ).rejects.toEqual("Please provide authedUser, qid, and answer");
  });
});

describe("snapshot test", () => {
  it("login", () => {
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
    );
    expect(component).toMatchSnapshot();
  });
});

describe("fireEvent test", () => {
  it("dashboard", () => {
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
    );
    const select = component.getByTestId("select");
    fireEvent.change(select, { target: { value: "sarahedo" } });
    const login = component.getByTestId("login");
    fireEvent.click(login);
    expect(component.getByTestId("avatar")).tobeinthedocument;
  });
});

describe("404 test", () => {
  it("checking message", () => {
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <PageNotFound />
        </BrowserRouter>
      </Provider>
    );
    const message = component.getByTestId("message");
    expect(message).tobeinthedocument;
  });
});

describe("dashboard test", () => {
  it("toggler", () => {
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
    );
    const toggler = component.getByTestId("toggler");
    expect(toggler).tobeinthedocument;
  });
});

describe("dashboard test", () => {
  it("navbar", () => {
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
    );
    const navbar = component.getByTestId("navbar");
    expect(navbar).tobeinthedocument;
  });
});

describe("create poll test", () => {
  it("form", () => {
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
    );
    const create_btn = component.getByTestId("create-btn");
    fireEvent.click(create_btn);
    expect(component.getByTestId("create-form")).tobeinthedocument;
  });
});
