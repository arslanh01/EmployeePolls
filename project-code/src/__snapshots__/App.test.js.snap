// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test login 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          style="direction: ltr;"
        >
          <div
            class=""
            style="width: 0%; transition: width 800ms linear 0s; will-change: width, opacity; height: 3px; background-color: red; position: absolute; opacity: 0;"
          />
          <div
            style="display: table; clear: both;"
          />
        </div>
        <div
          class="center"
        >
          <img
            alt="login"
            src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcT3Mclb0NdAfReSwkqWDtxIh2Oc4vEyPMYzeg&usqp=CAU"
          />
          <br />
          <h3>
            Please login as a user to continue
          </h3>
          <br />
          <select
            data-testid="select"
          >
            <option
              value="None"
            />
            <option
              value="sarahedo"
            >
              sarahedo
            </option>
            <option
              value="tylermcginnis"
            >
              tylermcginnis
            </option>
            <option
              value="mtsamis"
            >
              mtsamis
            </option>
            <option
              value="zoshikanlu"
            >
              zoshikanlu
            </option>
          </select>
          <br />
          <br />
          <button
            data-testid="login"
            disabled=""
          >
            Login
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        style="direction: ltr;"
      >
        <div
          class=""
          style="width: 0%; transition: width 800ms linear 0s; will-change: width, opacity; height: 3px; background-color: red; position: absolute; opacity: 0;"
        />
        <div
          style="display: table; clear: both;"
        />
      </div>
      <div
        class="center"
      >
        <img
          alt="login"
          src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcT3Mclb0NdAfReSwkqWDtxIh2Oc4vEyPMYzeg&usqp=CAU"
        />
        <br />
        <h3>
          Please login as a user to continue
        </h3>
        <br />
        <select
          data-testid="select"
        >
          <option
            value="None"
          />
          <option
            value="sarahedo"
          >
            sarahedo
          </option>
          <option
            value="tylermcginnis"
          >
            tylermcginnis
          </option>
          <option
            value="mtsamis"
          >
            mtsamis
          </option>
          <option
            value="zoshikanlu"
          >
            zoshikanlu
          </option>
        </select>
        <br />
        <br />
        <button
          data-testid="login"
          disabled=""
        >
          Login
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
